File: app/src/main/java/com/blues/application/BluesApplication.java
package com.blues.application;

import android.app.Activity;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkRequest;
import android.os.Build;
import android.os.Bundle;

import com.blues.framework.network.NetworkCallbackImpl;
import com.blues.framework.utils.ActivityManagerUtil;
import com.blues.framework.utils.Utils;

import androidx.multidex.MultiDex;
import androidx.multidex.MultiDexApplication;


/**
 * User : Blues
 * Date : 2019/3/6
 * Email : huajianlan@rastar.com
 */

public class BluesApplication extends MultiDexApplication {

//    private RefWatcher refWatcher;

    private Context mActivity;

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        //当引入第三方jar包过多时，尽心dex分包处理
        MultiDex.install(this);

    }

    @Override
    public void onCreate() {
        super.onCreate();

//        refWatcher = setupLeakCanary();

        //初始化工具类
        Utils.init(this);

        //初始化数据库
//        ThreadManager.getInstance().executeRunnable(() -> {
//            CourseManager.getInstance().initDataBase(getApplicationContext());
//        });

        //注册一下网络监听
        initNetworkListener();


        //批量管理activity
        registerActivityLifecycleCallbacks(new ActivityLifecycleCallbacks() {
            @Override
            public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
                ActivityManagerUtil.activityList.add(activity);
            }

            @Override
            public void onActivityStarted(Activity activity) {

            }

            @Override
            public void onActivityResumed(Activity activity) {

            }

            @Override
            public void onActivityPaused(Activity activity) {

            }

            @Override
            public void onActivityStopped(Activity activity) {

            }

            @Override
            public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

            }

            @Override
            public void onActivityDestroyed(Activity activity) {
                ActivityManagerUtil.activityList.remove(activity);
            }
        });
    }

    private void initNetworkListener() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
            ConnectivityManager.NetworkCallback callback = new NetworkCallbackImpl(mActivity);
            NetworkRequest.Builder builder = new NetworkRequest.Builder();
            NetworkRequest request = builder.build();
            ConnectivityManager manager = (ConnectivityManager) this.getSystemService(CONNECTIVITY_SERVICE);
            if (manager != null) {
                manager.registerNetworkCallback(request, callback);
            }
        }
    }

//    private RefWatcher setupLeakCanary() {
//
//        if (LeakCanary.isInAnalyzerProcess(this)) {
//            return RefWatcher.DISABLED;
//        }
//        return LeakCanary.install(this);
//    }

    //使用leak canary
//    public static RefWatcher getRefWatcher(Context context) {
//        BluesApplication application = (BluesApplication) context.getApplicationContext();
//        return application.refWatcher;
//    }
}

File: app/src/main/java/com/blues/application/BluesApplication.kt
package com.blues.application

import ActivityUtil
import androidx.multidex.MultiDexApplication
import androidx.multidex.MultiDex
import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.os.Build.VERSION.SDK_INT
import coil.Coil
import coil.ImageLoader
import coil.decode.GifDecoder
import coil.decode.ImageDecoderDecoder
import com.blues.adapter.ActivityLifecycleCallbacksAdapter
import com.blues.di.allModules
import com.blues.framework.http.OkHttpClientManager
import com.blues.framework.network.NetworkCallbackImpl
import com.tencent.mmkv.MMKV
import org.koin.android.ext.koin.androidContext
import org.koin.android.ext.koin.androidLogger
import org.koin.core.context.startKoin

/**
 * User : Blues
 * Date : 2019/3/6
 * Email : huajianlan@rastar.com
 */
class BluesApplication : MultiDexApplication() {

    companion object {

        lateinit var app: Context
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base) //当引入第三方jar包过多时，尽心dex分包处理
        MultiDex.install(this)
        app = base
    }

    override fun onCreate() {
        super.onCreate()

        //注册一下网络监听
        initNetworkListener()

        //初始化Activity声明周期监听
        initActivityLifecycleListener()

        //初始化koin
        initKoin()

        //初始化coil
        initCoil()

        //注册mmkv
        MMKV.initialize(this)
    }

    private fun initNetworkListener() {
        NetworkCallbackImpl(app)
    }

    private fun initKoin() {
        startKoin {
            androidContext(this@BluesApplication)
            androidLogger()
            modules(allModules)
        }
    }

    private fun initActivityLifecycleListener() {
        //批量管理activity
        registerActivityLifecycleCallbacks(object : ActivityLifecycleCallbacksAdapter() {

            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
                ActivityUtil.activities.add(activity)
            }

            override fun onActivityDestroyed(activity: Activity) {
                ActivityUtil.activities.remove(activity)
            }
        })
    }

    /**
     * 初始化coil
     */
    private fun initCoil() {
        ImageLoader.Builder(applicationContext).crossfade(true).componentRegistry {
                if (SDK_INT >= 28) {
                    add(ImageDecoderDecoder(applicationContext))
                } else {
                    add(GifDecoder())
                }
            }.okHttpClient {
                OkHttpClientManager.instance
            }.build().also {
                Coil.setImageLoader(it)
            }
    }
}